piptools/utils.py

```diff
@@ -522,3 +531,126 @@ def parse_requirements_from_wheel_metadata(
+    will use the first file found, searching in this order: an explicitly given config file, a
+    `.pip-tools.toml`, a `pyproject.toml` file. Those files are searched for in the same directory
+    as the requirements input file.
+    """
+    if value is None:
+        config_file = select_config_file(ctx.params.get("src_files", ()))
+        if config_file is None:
+            return None
+    else:
+        config_file = Path(value)
+
+    try:
+        config = parse_config_file(config_file)
+    except OSError as e:
+        raise click.FileError(
+            filename=str(config_file), hint=f"Could not read '{config_file}': {e}"
+        )
+    except ValueError as e:
+        raise click.FileError(
+            filename=str(config_file), hint=f"Could not parse '{config_file}': {e}"
+        )
+
+    if not config:
+        return None
+
+    defaults: dict[str, Any] = ctx.default_map.copy() if ctx.default_map else {}
+    defaults.update(config)
+
+    ctx.default_map = defaults
+    return str(config_file)
+
+
+def select_config_file(src_files: tuple[str, ...]) -> Path | None:
+    """
+    Returns the config file to use for defaults given `src_files` provided.
+    """
+    if not src_files:
+        # If no src_files were specified, we consider the current directory the only candidate
+        candidate_dirs = [Path.cwd()]
+    else:
+        # Collect the candidate directories based on the src_file arguments provided
+        src_files_as_paths = [
+            Path(Path.cwd(), src_file).resolve() for src_file in src_files
+        ]
+        candidate_dirs = [
+            src if src.is_dir() else src.parent for src in src_files_as_paths
+        ]
+    config_file_path = next(
+        (
+            candidate_dir / config_file
+            for candidate_dir in candidate_dirs
+            for config_file in (CONFIG_FILE_NAME, "pyproject.toml")
+            if (candidate_dir / config_file).is_file()
+        ),
+        None,
+    )
+    return config_file_path
+
+
+# Some of the defined click options have different `dest` values than the defaults
+NON_STANDARD_OPTION_DEST_MAP: dict[str, str] = {
+    "extra": "extras",
+    "upgrade_package": "upgrade_packages",
+    "resolver": "resolver_name",
+    "user": "user_only",
+}
+
+
+def get_click_dest_for_option(option_name: str) -> str:
+    """Returns the click `dest` value for the given option name."""
+    # Format the keys properly
+    option_name = option_name.lstrip("-").replace("-", "_").lower()
+    # Some options have dest values that are overrides from the click generated default
+    option_name = NON_STANDARD_OPTION_DEST_MAP.get(option_name, option_name)
+    return option_name
+
+
+# Ensure that any default overrides for these click options are lists, supporting multiple values
+MULTIPLE_VALUE_OPTIONS = [
+    "extras",
+    "upgrade_packages",
+    "unsafe_package",
+    "find_links",
+    "extra_index_url",
+    "trusted_host",
+]
+
+
+@functools.lru_cache()
```

Comment:
nit: does this need to be called if you don't pass any args?
```suggestion
@functools.lru_cache
```

Reply:
OTOH, maybe it makes sense to limit the cache size...
```suggestion
@functools.lru_cache(max_size=1)
```